### ISSUE TITLE
Question about the behavior of the seq_kd flag: does it actually implement Sequence-Level KD?

### ISSUE THREAD

**Issue by Tianyi-Lin** [shawnli/on-policy-distillation#1]
Hello, thank you very much for releasing this reproduction! I really appreciate the clarity of the codebase, and it has been extremely helpful for my research.

After reading the ICLR 2024 paper, my understanding of Sequence-Level Knowledge Distillation (SeqKD) is as follows:

Given a prompt, the teacher model generates a full answer sequence, and the student model is trained with NLL on that teacher-generated sequence. In other words, SeqKD performs supervised fine-tuning on teacher-generated outputs (off-policy).

However, when looking at the implementation in this repository, I noticed that seq_kd is used together with use_on_policy inside the following logic:
 

> if use_on_policy and self.config.seq_kd:
>     # On-policy: student model generates sequences
>     generated_ids, generated_mask = self.generate_on_policy_sequences(batch["input_ids"], batch["attention_mask"])
>     # Use the generated sequences as training data
>     input_ids = generated_ids
>     attention_mask = generated_mask
>     labels = generated_ids.clone()
> 
> else:
>     # Off-policy: use dataset sequences
>     input_ids = batch["input_ids"]
>     attention_mask = batch["attention_mask"]
>     labels = batch.get("labels", input_ids.clone())

This behavior seems to indicate that:
	•	When seq_kd + use_on_policy are both enabled, the student model generates the training sequences,
	•	And no teacher-generated sequences are involved.

This appears quite different from the SeqKD definition in the paper, where sequences should be generated by the teacher.

My Questions
	1.	Is seq_kd functioning merely as a switch for enabling on-policy distillation?
It does not seem to trigger any teacher-generation step as described in the paper.
	2.	Is the current implementation intended to represent the SeqKD method from the ICLR 2024 paper?
In the paper, SeqKD requires teacher-generated sequences, but this logic exclusively uses student-generated sequences.
	3.	If the intent is to use seq_kd as an on-policy KD flag, could this naming be misleading?
Since it differs from the conventional definition of SeqKD (teacher-generated outputs used as labels).

I would greatly appreciate clarification on:
	•	The actual purpose of the seq_kd flag in this codebase
	•	Whether teacher-generated sequences are implemented elsewhere
	•	Whether this design intentionally departs from the paper’s SeqKD description

Thank you again for the excellent work and for taking the time to look into this question! I really appreciate your help.
